--- /home/cmeng/source/smack/Smack-4.4.0-alpha3-20200404/org/jivesoftware/smack/AbstractXMPPConnection.java	2020-04-04 13:03:28.000000000 
+++ /home/cmeng/workspace/android/atalk-android/aTalk/src/main/java/org/jivesoftware/smack/AbstractXMPPConnection.java	2020-04-06 08:35:44.000000000 
@@ -248,13 +248,13 @@
             new HashMap<>();
 
     private final Map<Consumer<MessageBuilder>, GenericInterceptorWrapper<MessageBuilder, Message>> messageInterceptors = new HashMap<>();
 
     private final Map<Consumer<PresenceBuilder>, GenericInterceptorWrapper<PresenceBuilder, Presence>> presenceInterceptors = new HashMap<>();
 
-    private XmlEnvironment incomingStreamXmlEnvironment;
+    public XmlEnvironment incomingStreamXmlEnvironment;
 
     protected XmlEnvironment outgoingStreamXmlEnvironment;
 
     final MultiMap<QName, NonzaCallback> nonzaCallbacksMap = new MultiMap<>();
 
     protected final Lock connectionLock = new ReentrantLock();
