--- /home/cmeng/source/smack/Smack/trunk/smack-extensions/src/main/java/org/jivesoftware/smackx/bytestreams/ibb/InBandBytestreamManager.java	2018-11-30 05:55:38.000000000 
+++ /home/cmeng/workspace/android/smack/smack-extensions/src/main/java/org/jivesoftware/smackx/bytestreams/ibb/InBandBytestreamManager.java	2018-12-17 08:57:39.000000000 
@@ -297,12 +297,25 @@
      */
     public void ignoreBytestreamRequestOnce(String sessionID) {
         this.ignoredBytestreamRequests.add(sessionID);
     }
 
     /**
+     * Use this method to clear the next incoming In-Band Bytestream request containing the given
+     * session ID.
+     * <p>
+     * This method should be used if you are awaiting an In-Band Bytestream request as a reply to
+     * another stanza (e.g. file transfer).
+     *
+     * @param sessionID to be be cleared
+     */
+    public void removeIgnoredBytestreamRequest(String sessionID) {
+        this.ignoredBytestreamRequests.remove(sessionID);
+    }
+
+    /**
      * Returns the default block size that is used for all outgoing in-band bytestreams for this
      * connection.
      * <p>
      * The recommended default block size is 4096 bytes. See <a
      * href="http://xmpp.org/extensions/xep-0047.html#usage">XEP-0047</a> Section 5.
      *
