--- /home/cmeng/source/smack/Smack-4.4.0-alpha3-20200404/org/jivesoftware/smackx/bytestreams/ibb/InBandBytestreamSession.java	2020-03-24 09:16:12.000000000 
+++ /home/cmeng/workspace/android/atalk-android/aTalk/src/main/java/org/jivesoftware/smackx/bytestreams/ibb/InBandBytestreamSession.java	2020-04-06 09:29:14.000000000 
@@ -214,24 +214,27 @@
             }
         }
 
         if (this.inputStream.isClosed && this.outputStream.isClosed) {
             this.isClosed = true;
 
-            // send close request
-            Close close = new Close(this.byteStreamRequest.getSessionID());
-            close.setTo(this.remoteJID);
-            try {
-                connection.createStanzaCollectorAndSend(close).nextResultOrThrow();
-            }
-            catch (Exception e) {
-                // Sadly we are unable to use the IOException(Throwable) constructor because this
-                // constructor is only supported from Android API 9 on.
-                IOException ioException = new IOException();
-                ioException.initCause(e);
-                throw ioException;
+            // Do not send close stream IQ if on receive, otherwiese XMPPError: item-not-found - cancel
+            if (!in) {
+                // send close request
+                Close close = new Close(this.byteStreamRequest.getSessionID());
+                close.setTo(this.remoteJID);
+                try {
+                    connection.createStanzaCollectorAndSend(close).nextResultOrThrow();
+                }
+                catch (Exception e) {
+                    // Sadly we are unable to use the IOException(Throwable) constructor because this
+                    // constructor is only supported from Android API 9 on.
+                    IOException ioException = new IOException();
+                    ioException.initCause(e);
+                    throw ioException;
+                }
             }
 
             this.inputStream.cleanup();
 
             // remove session from manager
             // Thanks Google Error Prone for finding the bug where remove() was called with 'this' as argument. Changed
