--- org/jivesoftware/smack/provider/Provider.java	2019-06-03 23:41:49.000000000
+++ org/jivesoftware/smack/provider/Provider.java	2019-07-29 10:35:34.000000000
@@ -52,12 +52,15 @@
             currentType = currentClass.getGenericSuperclass();
         }
         ParameterizedType parameterizedGenericSuperclass = (ParameterizedType) currentType;
         Type[] actualTypeArguments = parameterizedGenericSuperclass.getActualTypeArguments();
         Type elementType = actualTypeArguments[0];
 
+        if (!(elementType instanceof Class))
+            elementType = elementType.getClass();
+
         elementClass =  (Class<E>) elementType;
     }
 
     public final Class<E> getElementClass() {
         return elementClass;
     }
