--- /home/cmeng/source/smack/Smack-4.4.0-alpha3-20200404/org/jivesoftware/smack/util/StringUtils.java	2020-04-04 13:03:28.000000000 
+++ /home/cmeng/workspace/android/atalk-android/aTalk/src/main/java/org/jivesoftware/smack/util/StringUtils.java	2020-04-05 15:21:22.000000000 
@@ -258,12 +258,30 @@
             hexChars[j * 2] = HEX_CHARS[v >>> 4];
             hexChars[j * 2 + 1] = HEX_CHARS[v & 0x0F];
         }
         return new String(hexChars);
     }
 
+    /**
+     * Convert a hexadecimal String to bytes.
+     *
+     * Source: https://stackoverflow.com/a/140861/11150851
+     *
+     * @param s hex string
+     * @return byte array
+     */
+    public static byte[] hexStringToByteArray(String s) {
+        int len = s.length();
+        byte[] data = new byte[len / 2];
+        for (int i = 0; i < len; i += 2) {
+            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
+                    + Character.digit(s.charAt(i + 1), 16));
+        }
+        return data;
+    }
+
     public static byte[] toUtf8Bytes(String string) {
         return string.getBytes(StandardCharsets.UTF_8);
     }
 
     /**
      * 24 upper case characters from the latin alphabet and numbers without '0' and 'O'.
