--- a/smack/smack-extensions/src/main/java/org/jivesoftware/smackx/xhtmlim/XHTMLText.java	2018-08-02 21:29:59.000000000
+++ b/smack/smack-extensions/src/main/java/org/jivesoftware/smackx/xhtmlim/XHTMLText.java	2018-08-02 21:29:59.000000000
@@ -102,13 +102,13 @@
      * @param lang the language of the body
      * @return this.
      */
     private XHTMLText appendOpenBodyTag(String style, String lang) {
         text.halfOpenElement(Message.BODY);
         text.xmlnsAttribute(NAMESPACE);
-        text.optElement(STYLE, style);
+        text.optAttribute(STYLE, style);
         text.xmllangAttribute(lang);
         text.rightAngleBracket();
         return this;
     }
 
     /**
