--- /home/cmeng/source/smack/Smack/trunk/smack-extensions/src/main/java/org/jivesoftware/smackx/filetransfer/FaultTolerantNegotiator.java	2017-12-25 21:16:43.000000000 
+++ /home/cmeng/workspace/android/smack/smack-extensions/src/main/java/org/jivesoftware/smackx/filetransfer/FaultTolerantNegotiator.java	2018-12-16 19:58:03.000000000 
@@ -62,17 +62,26 @@
                 "stream method.");
     }
 
     @Override
     public InputStream createIncomingStream(final StreamInitiation initiation) throws SmackException, XMPPErrorException, InterruptedException {
         // This could be either an xep47 ibb 'open' iq or an xep65 streamhost iq
+        InputStream inputStream = null;
         IQ initiationSet = initiateIncomingStream(connection(), initiation);
 
         StreamNegotiator streamNegotiator = determineNegotiator(initiationSet);
 
-        return streamNegotiator.negotiateIncomingStream(initiationSet);
+        try {
+            inputStream = streamNegotiator.negotiateIncomingStream(initiationSet);
+        } catch (Exception ex) {
+            if ((streamNegotiator instanceof Socks5TransferNegotiator)
+                    && (secondaryNegotiator instanceof IBBTransferNegotiator)) {
+                inputStream = ((IBBTransferNegotiator) secondaryNegotiator).getIbbIncomingStream(initiation);
+            }
+        }
+        return inputStream;
     }
 
     private StreamNegotiator determineNegotiator(Stanza streamInitiation) {
         if (streamInitiation instanceof Bytestream) {
             return primaryNegotiator;
         } else if (streamInitiation instanceof Open) {
