--- a/ffmpeg/libavcodec/hevc_mvs.c
+++ b/ffmpeg/libavcodec/hevc_mvs.c
@@ -20,12 +20,13 @@
  * License along with FFmpeg; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
 
 #include "hevc.h"
 #include "hevcdec.h"
+#undef B0
 
 static const uint8_t l0_l1_cand_idx[12][2] = {
     { 0, 1, },
     { 1, 0, },
     { 0, 2, },
     { 2, 0, },
